### ISSUE TITLE
Support for ARM 64 (Apple Silicon)

### ISSUE THREAD

**Issue by Francyrad** [hannorein/rebound#681]
Dear developers
I've been able to successfully run some example that didn't require the use of MPI.
Anyway, when it comes to install some simulations that require mpi (in thos case "self_gravity_disc_mpi), it's impossible to compile the program. I attach here the error, which should be given by some incompatibilities with arm64 architecture. I hope you can solve
Francesco.
[rebound_error_mpi.txt](https://github.com/hannorein/rebound/files/11216640/rebound_error_mpi.txt)

**Comment by hannorein** [comment]
The errors are not related to ARM. It's just that the MPI part of REBOUND hasn't been keeping up with a lot of changes. This is mainly due to me not having a good problem where MPI might be helpful (OpenMP is often enough). It would take some effort to get MPI working again, and the detail will depend a lot on the specific problem you're trying to solve.

**Comment by Francyrad** [comment]
Thank you for your response
Actually Rebound would be the only program that could help me. I want to simulate the formation of the moon starting from a disk and around the proto earth like in this article where the authors used Rebound:

"FORMATION OF MULTIPLE-SATELLITE SYSTEMS FROM LOW-MASS CIRCUMPLANETARY PARTICLE DISKS"

To account the particle interactions when my moon will be formed, i will need all my processors using mpi.
Actually i'm using a very good program called OpenSPH that is able to do that works, but when aggregation happen the code can only run in 1 core, so my simulation toke 3 months to finish and with only 10k particles. So i'd like to test Rebound to run much more particles.

I hope you can help me to solve this problem, and letting me know if I really need MPI to solve this problem.

Thank you for your help
Francesco

**Comment by hannorein** [comment]
10k is not very large. You might not get the performance from MPI that you're hoping for. I'd recommend using OpenMP for shared memory parallelization. Most computing clusters will have nodes that can run tens of threads in parallel. With that you effectively have a few hundred particles per core. MPI on top of that will most likely not help.

**Comment by Francyrad** [comment]
Yes, it is not very large, but it is when i must run that only in 1 core with open. Will OpenMP help me to run the simulation with 100k - 1kk particles or more?

**Comment by hannorein** [comment]
I can't give you a straight answer. You need to try it out for your specific setup.

**Comment by Francyrad** [comment]
I will try to design the simulation and let you know. thank you for your help!

**Comment by Francyrad** [comment]
Dear professor @hannorein 
I've successfully designed my simulation and it runs... 

<img width="699" alt="Screenshot 2023-04-28 alle 14 48 23" src="https://user-images.githubusercontent.com/51544149/235154012-4464b57e-1514-4f09-8a81-61dbaa736cd8.png">


except for the detail that OpenMP doesn't work for some reason and it runs only in 1 core:

<img width="845" alt="Screenshot 2023-04-28 alle 14 59 04" src="https://user-images.githubusercontent.com/51544149/235154111-fdd5d79a-032d-4e69-9ee1-dfda6833a425.png">

Changing the number of particles doesn't help, it became impossibly slow (terminal output):

```
N_tot=  10001     t=  2990.000000  dt=  1.000000  cpu=  3.597678 [s]  t/tmax=  0
N_tot=  10001     t=  3000.000000  dt=  1.000000  cpu=  3.649657 [s]  t/tmax=  0.00%

2023-04-28 14:47:38.844 rebound[20266:3701616] IMKClient Stall detected, *please Report* your user scenario attaching a spindump (or sysdiagnose) that captures the problem - (imkxpc_bundleIdentifierWithReply:) block performed very slowly (2.49 secs).

N_tot=  10001     t=  3010.000000  dt=  1.000000  cpu=  3.675214 [s]  t/tmax=  0
N_tot=  10001     t=  3020.000000  dt=  1.000000  cpu=  3.625312 [s]  t/tmax=  0
N_tot=  10001     t=  3030.000000  dt=  1.000000  cpu=  3.659524 [s]  t/tmax=  0
N_tot=  10001     t=  3040.000000  dt=  1.000000  cpu=  3.607195 [s]  t/tmax=  0
N_tot=  10001     t=  3050.000000  dt=  1.000000  cpu=  3.600862 [s]  t/tmax=  0
N_tot=  10001     t=  3060.000000  dt=  1.000000  cpu=  3.610075 [s]  t/tmax=  0
N_tot=  10001     t=  3070.000000  dt=  1.000000  cpu=  3.608410 [s]  t/tmax=  0
N_tot=  10001     t=  3080.000000  dt=  1.000000  cpu=  3.590928 [s]  t/tmax=  0
N_tot=  10001     t=  3090.000000  dt=  1.000000  cpu=  3.599406 [s]  t/tmax=  0
N_tot=  10001     t=  3100.000000  dt=  1.000000  cpu=  3.614358 [s]  t/tmax=  0
N_tot=  10001     t=  3110.000000  dt=  1.000000  cpu=  3.610601 [s]  t/tmax=  0
N_tot=  10001     t=  3120.000000  dt=  1.000000  cpu=  3.631834 [s]  t/tmax=  0
N_tot=  10001     t=  3130.000000  dt=  1.000000  cpu=  3.628529 [s]  t/tmax=  0
N_tot=  10001     t=  3140.000000  dt=  1.000000  cpu=  3.619742 [s]  t/tmax=  0
N_tot=  10001     t=  3150.000000  dt=  1.000000  cpu=  3.639600 [s]  t/tmax=  0
N_tot=  10001     t=  3160.000000  dt=  1.000000  cpu=  3.611394 [s]  t/tmax=  0.00
%zsh: terminated  ./rebound
francyrad@MacBook-Pro-di-Francesco moon % make
Compiling shared library librebound.so ...
/Library/Developer/CommandLineTools/usr/bin/make -C ../../src/
make[1]: Nothing to be done for `all'.
fatal: .git non è un repository Git (né lo è alcuna delle directory genitrici)

Compiling problem file ...
gcc -I../../src/ -Wl,-rpath,./ -std=c99 -Wpointer-arith -D_GNU_SOURCE -O3  -I/usr/local/include -Wall -g  -I/opt/homebrew/include -L/opt/homebrew/lib -fopenmp -D_APPLE -DOPENGL -DOPENMP -DGITHASH=0000000000gitnotfound0000000000000000000 problem.c -L. -lrebound -L/usr/local/lib -lglfw -framework Cocoa -framework OpenGL -framework IOKit -framework CoreVideo -fopenmp -o rebound

REBOUND compiled successfully.
francyrad@MacBook-Pro-di-Francesco moon % ./rebound
N_tot=  100001    t=  0.000000  dt=  1.000000  cpu=  0.000000 [s]  t/tmax=  0.00
N_tot=  100001    t=  10.000000  dt=  1.000000  cpu=  627.292950 [s]  t/tmax=  0.00%
```

In the user manual there is written that to enable OpenMP it is only necessary to compile the makefile correctly. I've MacOS Ventura with libomp installed with homebrew. This is the Makefile that i use:

```
export CC=gcc

ifeq ($(shell $(CC) -v 2>&1 | grep -c "clang"), 1)
export OPENMPCLANG=1
else
export OPENMP=1
endif

export OPENGL=1
include ../../src/Makefile.defs

all: librebound
	@echo ""
	@echo "Compiling problem file ..."
	$(CC) -I../../src/ -Wl,-rpath,./ $(OPT) $(PREDEF) problem.c -L. -lrebound $(LIB) -o rebound
	@echo ""
	@echo "REBOUND compiled successfully."

librebound: 
	@echo "Compiling shared library librebound.so ..."
	$(MAKE) -C ../../src/
	@-rm -f librebound.so
	@ln -s ../../src/librebound.so .

clean:
	@echo "Cleaning up shared library librebound.so ..."
	@-rm -f librebound.so
	$(MAKE) -C ../../src/ clean
	@echo "Cleaning up local directory ..."
	@-rm -vf rebound

```

It compiles with export OPENMP=1. When i try to force to compile with CLANG, the terminal gives to me the following error:

```
francyrad@MacBook-Pro-di-Francesco moon % make     
Compiling shared library librebound.so ...
/Library/Developer/CommandLineTools/usr/bin/make -C ../../src/
make[1]: Nothing to be done for `all'.
fatal: .git non è un repository Git (né lo è alcuna delle directory genitrici)

Compiling problem file ...
gcc -I../../src/ -Wl,-rpath,./ -std=c99 -Wpointer-arith -D_GNU_SOURCE -O3  -I/usr/local/include -Wall -g  -I/opt/homebrew/include -L/opt/homebrew/lib -I/include -Xpreprocessor -fopenmp -D_APPLE -DOPENGL -DOPENMP -DGITHASH=0000000000gitnotfound0000000000000000000 problem.c -L. -lrebound -L/usr/local/lib -lglfw -framework Cocoa -framework OpenGL -framework IOKit -framework CoreVideo -lomp -o rebound
ld: library not found for -lomp
collect2: error: ld returned 1 exit status
make: *** [all] Error 1
```

Is there some way how can I solve the problem? OpenMP is correctly installed and compiled with Homebrew. 
My processor is M1Pro with 10 cores.
Let me know and thank you in advance

Francesco

**Comment by hannorein** [comment]
I'm not sure. I've just tested it on an M1. If I compile and run `examples/openmp` without any changes, it works, for me. For that specific example (20k particles), I get a speed up of 4.5x on 8 cores. 

You probably have not correctly set up your environment variables (or alternatively the makefile). Specifically, you need to somewhere point the compiler to your OpenMP  header and library, e.g. 

```
 export LDFLAGS="-L/opt/homebrew/opt/libomp/lib"
 export CPPFLAGS="-I/opt/homebrew/opt/libomp/include"
```

Hanno

**Comment by Francyrad** [comment]
Thank you for your answer!
I've no idea what is going on honestly.
I already had these path in my bash (exactly the same lines) for some other program that I used (and that i do not remember honestly).

I runned the example as you did with 20K particles, this is my output:

```
francyrad@MacBook-Pro-di-Francesco openmp % make
Compiling shared library librebound.so ...
/Library/Developer/CommandLineTools/usr/bin/make -C ../../src/
make[1]: Nothing to be done for `all'.
fatal: .git non è un repository Git (né lo è alcuna delle directory genitrici)

Compiling problem file ...
gcc -I../../src/ -Wl,-rpath,./ -std=c99 -Wpointer-arith -D_GNU_SOURCE -O3  -I/usr/local/include -Wall -g  -fopenmp -D_APPLE -DOPENMP -DGITHASH=0000000000gitnotfound0000000000000000000 problem.c -L. -lrebound -L/usr/local/lib -fopenmp -o rebound

REBOUND compiled successfully.
francyrad@MacBook-Pro-di-Francesco openmp % ./rebound                      
2023-04-29 20:07:58.220 rebound[76316:5838142] IMKClient Stall detected, *please Report* your user scenario attaching a spindump (or sysdiagnose) that captures the problem - (imkxpc_bundleIdentifierWithReply:) block performed very slowly (5.72 secs).


OpenMP speed-up: 0.999x (perfect scaling would give 10x)
```

I've no idea what is going on honestly, because the executable compile perfectly. I tried also to compile with clang and homebrew, the compile both, but the problem is the same.

This is also my (incomplete) code, I don't think it's fault of my code, but can you please check with the Macbook what does the output about OpenMP says? It would be of help. Thank you for your support.

```
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <math.h>
#include <time.h>
#include <sys/time.h>
#include <sys/stat.h>
#include <string.h>
#include <omp.h>
#include "rebound.h"
#include "tools.h"
#include "output.h"

void output_orbits(struct reb_simulation* const r);
void setup_simulation(struct reb_simulation* sim);
void heartbeat(struct reb_simulation* const r);
void output_energy(struct reb_simulation* const r);


const double G = 6.67430e-11;
const double M_earth = 5.972e24;
const double R_earth = 6371000.0;
const double rho_particle = 3300.0; // kg/m^3
const double r_inner = 6372000.0;
const double r_outer = 25484000.0;
const int N_particles = 1000;

const double M_moon = 7.342e22; // kg (massa lunare)
const double total_mass = M_moon; // massa totale delle particelle = 1 massa lunare

double random_uniform(double min, double max) {
    return min + (max - min) * ((double) rand() / (double) RAND_MAX);
}

void collision_resolve(struct reb_simulation* const sim, struct reb_collision c) {
    struct reb_particle* const particles = sim->particles;
    struct reb_particle p1 = particles[c.p1];
    struct reb_particle p2 = particles[c.p2];

    if (reb_collision_resolve_hardsphere(sim, c) == 0) {
        particles[c.p1].r = p1.r;
        particles[c.p2].r = p2.r;
    }
}

void output_angular_momentum(struct reb_simulation* const r);
void output_center_of_mass(struct reb_simulation* const r);

void setup_simulation(struct reb_simulation* sim) {
    sim->G = G;
    sim->integrator = REB_INTEGRATOR_LEAPFROG;
    sim->collision = REB_COLLISION_TREE;
    sim->collision_resolve = reb_collision_resolve_hardsphere;
    sim->boundary = REB_BOUNDARY_OPEN;
    sim->gravity = REB_GRAVITY_BASIC;
    sim->dt = 1;
    reb_configure_box(sim, 3 * r_outer, 3, 3, 3);
    sim->softening = 0.02; // Gravitational softening length
    sim->opening_angle2 = 1.5; // This constant determines the accuracy of the tree code gravity estimate.

    struct reb_particle Earth = {0};
    Earth.m = M_earth;
    Earth.r = R_earth;
    reb_add(sim, Earth);

    double M_particle = total_mass / N_particles;
    double R_particle = pow((3.0 * M_particle) / (4.0 * M_PI * rho_particle), 1.0 / 3.0) / 1000.0; // Raggio in km

    for (int i = 0; i < N_particles; i++) {
        double r = random_uniform(r_inner, r_outer);
        double theta = random_uniform(0, 2 * M_PI);

        struct reb_particle p = {0};
        p.m = M_particle;
        p.r = R_particle * 1000.0; // Raggio in metri
        p.x = r * cos(theta);
        p.y = r * sin(theta);
        p.z = 0;
        double v_kep = sqrt(G * M_earth / r);
                p.vx = -v_kep * sin(theta);
                p.vy = v_kep * cos(theta);
                p.vz = 0;

                reb_add(sim, p);
            }
        }

        int main() {
            struct reb_simulation* sim = reb_create_simulation();
            setup_simulation(sim);
            sim->heartbeat = heartbeat;

            FILE* f = fopen("particles.txt", "w");
            for (int i = 1; i < sim->N; i++) {
                fprintf(f, "%e %e %e\n", sim->particles[i].x, sim->particles[i].y, sim->particles[i].z);
            }
            fclose(f);

            //reb_simulationarchive_automate_interval(sim, "archive.bin", 10.);
            reb_integrate(sim, INFINITY);
            reb_free_simulation(sim);

            return 0;
        }

        void write_header(const char* filename, const char* header) {
            FILE* file = fopen(filename, "r");
            if (file == NULL) {
                file = fopen(filename, "w");
                fprintf(file, "%s\n", header);
                fclose(file);
            } else {
                fclose(file);
            }
        }

        void create_directory_if_not_exists(const char* dirname) {
            struct stat st = {0};
            if (stat(dirname, &st) == -1) {
                mkdir(dirname, 0700);
            }
        }

        void heartbeat(struct reb_simulation* const r) {
            double output_interval = 10.; // Intervallo di tempo tra gli output (in secondi)
            if (reb_output_check(r, output_interval)) {
                reb_output_timing(r, INFINITY);

                output_angular_momentum(r);
                output_center_of_mass(r);
                output_energy(r);
                //output_orbits(r);

                /*
                FILE* f_positions = fopen("output_data/output_positions.txt", "a");
                for (int i = 1; i < r->N; i++) {
                    fprintf(f_positions, "%e %e %e\n", r->particles[i].x, r->particles[i].y, r->particles[i].z);
                }
                fprintf(f_positions, "\n"); // Aggiunge una riga vuota tra gli output per facilitare la lettura
                fclose(f_positions);
                */
            }
        }

        void output_angular_momentum(struct reb_simulation* const r) {
            const double t = r->t;
            const int N = r->N;
            struct reb_particle* const particles = r->particles;

            double L_tot = 0.0;
            double Lx = 0.0;
            double Ly = 0.0;
            double Lz = 0.0;

            for (int i = 0; i < N; i++) {
                struct reb_particle p = particles[i];
                Lx += p.m * (p.y * p.vz - p.z * p.vy);
                Ly += p.m * (p.z * p.vx - p.x * p.vz);
                Lz += p.m * (p.x * p.vy - p.y * p.vx);
            }
            L_tot = sqrt(Lx * Lx + Ly * Ly + Lz * Lz);

            create_directory_if_not_exists("output_data");
            write_header("output_data/angular_momentum.txt", "Time\tang_mom_x\tang_mom_y\tang_mom_z\ttot_ang_mom");
            FILE* f = fopen("output_data/angular_momentum.txt", "a");
            fprintf(f, "%.16e\t%.16e\t%.16e\t%.16e\t%.16e\n", t, Lx, Ly, Lz, L_tot);
                    fclose(f);
            }

            void output_center_of_mass(struct reb_simulation* const r) {
                const double t = r->t;
                const int N = r->N;
                struct reb_particle* const particles = r->particles;

                double M_tot = 0.0;
                double com_x = 0.0;
                double com_y = 0.0;
                double com_z = 0.0;

                for (int i = 0; i < N; i++) {
                    struct reb_particle p = particles[i];
                    M_tot += p.m;
                    com_x += p.m * p.x;
                    com_y += p.m * p.y;
                    com_z += p.m * p.z;
                }
                com_x /= M_tot;
                com_y /= M_tot;
                com_z /= M_tot;

                create_directory_if_not_exists("output_data");
                write_header("output_data/center_of_mass.txt", "Time\tcom_x\tcom_y\tcom_z\ttot_com");
                FILE* f = fopen("output_data/center_of_mass.txt", "a");
                double tot_com = sqrt(com_x * com_x + com_y * com_y + com_z * com_z);
                fprintf(f, "%.16e\t%.16e\t%.16e\t%.16e\t%.16e\n", t, com_x, com_y, com_z, tot_com);
                fclose(f);
            }

            void write_orbits_header(const char* filename, const char* header) {
                FILE* file = fopen(filename, "r");
                if (file == NULL) {
                    file = fopen(filename, "w");
                    fprintf(file, "%s\n", header);
                    fclose(file);
                } else {
                    fclose(file);
                }
            }

            void output_orbits(struct reb_simulation* const r) {
                create_directory_if_not_exists("output_data");
                write_orbits_header("output_data/orbits.txt",
                                    "time\tsemi-major_axis\teccentricity\tinclination\tOmega\tomega\tlambda\torbital_period\ttrue_anomaly");
                reb_output_orbits(r, "output_data/orbits.txt");
            }


            void output_energy(struct reb_simulation* const r) {
                double total_energy = reb_tools_energy(r);
                double t = r->t;

                create_directory_if_not_exists("output_data");
                write_header("output_data/energy.txt", "Time\tTotal_Energy");
                FILE* f = fopen("output_data/energy.txt", "a");
                fprintf(f, "%.16e\t%.16e\n", t, total_energy);
                fclose(f);
            }
```

**Comment by hannorein** [comment]
I can't run your entire code for you. It's too long (and you call it incomplete). Try to simplify your code. Take out everything that's not important to narrow down the problem. Compare it with the example. Figure out if any difference are making a difference. Note that not all parts in REBOUND are parallelized.

**Comment by Francyrad** [comment]
Sure, i'll try to take out all the non necessary part and i'll do a comparision with the example

**Comment by Francyrad** [comment]
I tired the example many times and only once i got a scaling of 2.7.
openMP works, but i don't understand why i can't get the scaling, and i'm not running other operations in background... What could be the problem?
This is why i'd love to have mpi for rebound, because it will be sure that i can use all my processors... At least this is my personal case that i encountered with other programs

**Comment by hannorein** [comment]
Compare it with the example that works. Figure out what the differences are. Note that not all parts in REBOUND are parallelized.

**Comment by Francyrad** [comment]
I tried to run the code with my old intel computer and it is able to scale, also my code. I did some research and, as I can see, it's an ARM64 with some chips problem that do not let you scale OpenMP, also other people met this problem...

The chips involved are:  M1Pro (mine), M1Ultra, M1MAX. Only M1 avoid the issue because the chip is older, this is why you got the scaling.

unless OpenMP developers fix this, it is not possible to run the code in parallel. This is why i'd love MPI support for rebound in the future. In my experience i never met problem with other programs.

Thank you professor for your support, i'll close the isse when the OpenMP problem is solved
Francesco
